Cannabis has more believable effects

cacogen opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

- Adds cannabis reagent and replaces space drugs with it in cannabis products
- Adds space drugs to rainbow weed, which fits its theme and might satisfy the people who miss the effect
- Removes lipolicide from cannabis products in favour of the reagent consuming nutrition directly
- Cannabis reagent gives you messages describing symptoms of cannabis intoxication, makes you laugh, makes you hungry, has a chance to couchlock you buckled to a chair and put you to sleep lying down
- Cannabis no longer references hallucinating, can't be overdosed and has no addiction
- Adds Stoned status effect, which is applied by cannabis reagent
- Stoned status effect slows you down, gives you bloodshot eyes, makes you clumsy, boosts your mood and gives you the drugged sound environment
- Adds "stoned" icon state for Stoned status effect screen alert (sprite shamelessly stolen from the Smoke spell)
- Adds BLOODSHOT_EYES trait, which is applied by Stoned status effect
- Adds examine eye descriptions for bloodshot eyes and player-chosen red eyes
- Adds cannabis movespeed modifier, which is applied by the Stoned status effect
- Adds Stoned mood event, which is applied by Stoned status effect and gives the same mood boost as High but doesn't reference tripping
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

Makes the drug more true to life.
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl: cacogen
balance: Cannabis has more believable effects
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

cacogen committed:
Cannabis makes your eyes red

- Adds cannabis reagent and replaces space drugs with it in all cannabis products
- Cannabis reagent metabolises at the same rate as nicotine, gives you messages describing symptoms of cannabis intoxication, makes you laugh and makes you hungry
- Cannabis products no longer have hallucinogenic properties, can't be overdosed on and are non-addictive (which makes them more believable)
- Removes lipolicide from cannabis products in favour of the reagent consuming nutrition directly
- Adds Stoned status effect, which is applied by Cannabis reagent
- Stoned status effect slows you down, gives you bloodshot eyes, makes you clumsy, boosts your mood and gives you the drugged sound environment
- Adds stoned icon state for Stoned status effect screen alert (shamelessly stolen from Smoke spell)
- Adds BLOODSHOT_EYES trait, which is applied by Stoned status effect
- Adds examine eye descriptions for bloodshot eyes and player-chosen red eyes
- Adds Cannabis movespeed modifier, which is applied by the Stoned status effect
- Adds Stoned mood event, which is applied by Stoned status effect and is basically the same as High but doesn't reference tripping

maxymax13 commented:
finally, i made weed in botany the other day and it just made me vomit and made the screen change colors like a fifth grader's imagination of weed after watching family guy

maxymax13 commented on code/datums/status_effects/drug_effects.dm:
```diff
@@ -58,3 +58,39 @@
+
+/atom/movable/screen/alert/status_effect/stoned
+	name = "Stoned"
+	desc = "nigga u high as shit"
```
😕

maxymax13 commented on code/modules/mob/living/carbon/human/examine.dm:
```diff
@@ -83,6 +83,10 @@
+		else if(eye_color == BLOODCULT_EYE && HAS_TRAIT(src, BLOODSHOT_EYES))
+			. += "<span class='warning'><B>[t_His] eyes are bloodshot!</B></span>"
+		else if(eye_color == BLOODCULT_EYE)
+			. += "<span class='warning'><B>[t_His] eyes are an unnatural red!</B></span>"
```
what a weird distinction to make

cacogen replied:
nobody chooses to be a cultist

maxymax13 commented on code/modules/movespeed/modifiers/reagent.dm:
```diff
@@ -12,8 +12,7 @@
 	multiplicative_slowdown = -0.35
-
-/datum/movespeed_modifier/reagent/changelinghaste
+datum/movespeed_modifier/reagent/changelinghaste
```
bruh

maxymax13 commented on code/modules/reagents/chemistry/reagents/drug_reagents.dm:
```diff
@@ -34,6 +34,26 @@
+		to_chat(M, "<span class='notice'>[smoke_message]</span>")
+	if(DT_PROB(1, delta_time))
+		M.emote(pick("chuckle","smile","laugh","giggle"))
+	M.adjust_nutrition(-0.5 * REM * delta_time) //munchies
```
hell fucking yeah

Mickyan commented:
Removing space drugs in favor of real ones is absolutely lame and we should do the exact opposite

maxymax13 commented:
> 
> 
> Removing space drugs in favor of real ones is absolutely lame and we should do the exact opposite

weed is already a thing you big egg

Mickyan commented:
It's space weed

not to mention cannabis is a plant and not a chemical

D4C-420 commented:
> not to mention cannabis is a plant and not a chemical

I'm pretty sure THC is the drug which we are talking about here so maybe call it that so it does not get confused with cannabis plants in code.

cacogen committed:
Fixes missing forward slash at the beginning of typepath

Co-authored-by: Kokonut <38844529+maxymax13@users.noreply.github.com>

maxymax13 commented:
if we're going to have weed at least make it make you act stoned instead of windows movie maker rainbow effect haha woah guys!

Mickyan commented:
my ideological reason is that this even more lame and boring for the sake of realism

interestingusernam3 commented on code/modules/reagents/chemistry/reagents/drug_reagents.dm:
```diff
@@ -34,6 +34,26 @@
 	..()
 
+/datum/reagent/drug/cannabis
+	name = "Cannabis"
```
```suggestion
	name = "Cannabis-derived cannabinoid mixture"
```

cacogen replied:
It's not technically accurate plus it isn't necessary. There are plenty of reagents that represent a group of chemicals or have vague names. Drinks, bath salts or space drugs for example.

maxymax13 commented on code/datums/status_effects/drug_effects.dm:
```diff
@@ -58,3 +58,39 @@
+
+/atom/movable/screen/alert/status_effect/stoned
+	name = "Stoned"
+	desc = "nigga u high as shit"
```
apparently, they resolved the other one

CPTANT commented:
Title doesn't really cover the contents of the PR.

MrDoomBringer commented:
> my ideological reason is that this even more lame and boring for the sake of realism

Just because the PR desc has the word 'realism' doesn't mean it's more "lame and boring"
This feature as it stands is way more interesting than the current affects of cannabis. Making "being high" just make your screen change colors is lame and boring compared to this, which actually implements real gameplay elements - IE the clumsy trait, having to consume way more food, etc.

Sec walking in on a stoner den and mistaking the huddled group of stumbling crew with red eyes for cultists is immensely funny and you cannot convince me otherwise

cacogen committed:
Saves the repo

Co-authored-by: interestingusernam3 <51925758+interestingusernam3@users.noreply.github.com>

tralezab commented on code/datums/status_effects/drug_effects.dm:
```diff
@@ -58,3 +58,39 @@
+
+/atom/movable/screen/alert/status_effect/stoned
+	name = "Stoned"
+	desc = "dood u high as shit"
```
literally how can anyone be against this pr

Mothblocks replied:
FJOIWEHUWQEFGYUWDGHUIWHUIDWEHUIFDUWGYSXQHUIODSDBNJKVBNKDML <--- this is you right now

cacogen replied:
That was the precedent that made think this would be okay also it is perfectly in tone with being high which should be kind of meta tbh

MMMiracles commented:
it should have a (high) OD threshold where it makes you fall asleep/be paralyzed temporarily from how absolutely baked you are

interestingusernam3 commented:
Looks like you got a spontaneous test fail, try pushing an empty commit.

cacogen commented:
>my ideological reason is that this even more lame and boring for the sake of realism

Truthfully I think it will be more boring to use without the screen effect. If I had something more fitting to replace it with I would. But other people can always add new effects later and I think some drugs having more milder effects than others is fine too.

cacogen committed:
Formats bloodshot_eyes define properly

- Adds space drugs to rainbow weed because it's fitting

cacogen committed:
Renames eye traits to have TRAIT_ prefix like others

- Cannabis can cause you to fall asleep lying down or couchlock you buckled to a chair

maxymax13 commented:
the screen effect is just boring and annoying, unless you find a way to get something cooler in BYOND i'd rather have nothing at all

cacogen commented:
I tried using blurriness but I think the effect is too pixelated to look good and too intrusive, especially in motion. I actually think not having a screen effect and being high being a more subtle effect actually helps it. It's like not realising how fucked up you are until you stand up. The echoey audio that I kept from space drugs helps it a lot.

@MMMiracles

Instead of making it an overdose effect as though having smoked too much (panic attack when) I chose to make it so being buckled upright has a chance to couchlock (paralyze) you for 10 seconds and lying down has a chance to make you fall asleep for 10 seconds. I'm not against overdose effects I just thought this was better.

carshalash commented:
Uhh, does not constantly yell "Dude weed lmao" not very authentic.

cacogen committed:
Removes sovl

cacogen commented on code/datums/status_effects/drug_effects.dm:
```diff
@@ -58,3 +58,39 @@
+		human_owner.eye_color = BLOODCULT_EYE //gives cultists plausible deniability
+		human_owner.update_body() //updates eye color
+		ADD_TRAIT(human_owner, TRAIT_BLOODSHOT_EYES, type) //dilates blood vessels in eyes
+		ADD_TRAIT(human_owner, TRAIT_CLUMSY, type) //impairs motor coordination
```
recently the image of a chemist with a smoke machine incapacitating an entire sec team with weed smoke came to mind and i'm not sure if it's going to be a problem

MrDoomBringer replied:
emergent gameplay mark as resolved 👍

Fikou commented on code/modules/mob/living/carbon/human/examine.dm:
```diff
@@ -81,8 +81,12 @@
 		if(glasses  && !(glasses.item_flags & EXAMINE_SKIP))
 			. += "[t_He] [t_has] [glasses.get_examine_string(user)] covering [t_his] eyes."
-		else if(eye_color == BLOODCULT_EYE && IS_CULTIST(src) && HAS_TRAIT(src, CULT_EYES))
+		else if(eye_color == BLOODCULT_EYE && IS_CULTIST(src) && HAS_TRAIT(src, TRAIT_CULT_EYES))
 			. += "<span class='warning'><B>[t_His] eyes are glowing an unnatural red!</B></span>"
+		else if(eye_color == BLOODCULT_EYE && HAS_TRAIT(src, TRAIT_BLOODSHOT_EYES))
+			. += "<span class='warning'><B>[t_His] eyes are bloodshot!</B></span>"
+		else if(eye_color == BLOODCULT_EYE)
+			. += "<span class='warning'><B>[t_His] eyes are an unnatural red!</B></span>"
```
the way youre doing this check is bad

Rohesie replied:
Instead of repeating the `eye_color == BLOODCULT_EYE` check three times, do it once and nest the rest inside of it:
```dm
else if(eye_color == BLOODCULT_EYE)
	if(S_CULTIST(src) && HAS_TRAIT(src, TRAIT_CULT_EYES))
		. += "<span class='warning'><B>[t_His] eyes are glowing an unnatural red!</B></span>"
	else if(HAS_TRAIT(src, TRAIT_BLOODSHOT_EYES))
		. += "<span class='warning'><B>[t_His] eyes are bloodshot!</B></span>"
```
I don't think you want the last one, as that'd change behavior.

Rohesie commented on code/modules/reagents/chemistry/reagents/drug_reagents.dm:
```diff
@@ -34,6 +34,32 @@
+		to_chat(M, "<span class='warning'>It's too comfy to move...</span>")
+		M.Paralyze(10 SECONDS)
+	..()
+	. = TRUE
```
Why does this require a mob health/stamina update?

cacogen replied:
uhhhhhhhhhhhhh

Rohesie replied:
Where did you take this from?

cacogen replied:
Probably nicotine below it

Rohesie replied:
The proc will force a health update if it returns a true value.
If there's no changes to stamina or health then that shouldn't be there.

Rohesie commented on code/datums/status_effects/drug_effects.dm:
```diff
@@ -58,3 +58,39 @@
+		var/mob/living/carbon/human/human_owner = owner
+		original_eye_color = human_owner.eye_color
+		human_owner.add_movespeed_modifier(/datum/movespeed_modifier/reagent/cannabis) //slows you down
+		human_owner.eye_color = BLOODCULT_EYE //gives cultists plausible deniability
```
Do they really have plausible deniability if the bloodshot trait is added?

cacogen replied:
Trying to recall my rationale. No, there's no plausible deniability unless the person examining them doesn't read the red eyes examine message. It makes a bit more ambiguous, though.

Rohesie replied:
Maybe change it so there is, or alter the comment, then.

Rohesie commented on code/modules/reagents/chemistry/reagents/drug_reagents.dm:
```diff
@@ -34,6 +34,32 @@
+	metabolization_rate = 0.125 * REAGENTS_METABOLISM
+
+/datum/reagent/drug/cannabis/on_mob_life(mob/living/carbon/M, delta_time, times_fired)
+	M.apply_status_effect(/datum/status_effect/stoned)
```
This shouldn't be re-applied on every mob life cycle.
Make it a status effect without maximum duration and apply it on first metabolization, removing it when the chem leaves the body.
Make sure the metabolization rate is slow.

cacogen replied:
It took me so many hours to get the red eyes to work. I didn't use a status effect originally, but oranges told me a timed status effect would get it to work and it did. I tried different metabolisation rates (such as the one used by meth, which applies a movement speed modifier), but anytime the reagent added the red eyes they would immediately disappear again. I'm guessing the same thing will happen with the status effect. I'll try it, though. I don't know a lot about how reagent metabolisation works.

Rohesie replied:
The status effect is fine. Just add it on first metabolisation and remove it on last, and make sure the status effect does not have a set duration, but that it stays there until removed.

cacogen replied:
I'll try it but what I'm saying is I don't think it will work if the status effect isn't timed. It will just disappear immediately, like with the red eyes trait.

cacogen replied:
Yeah it doesn't work

Rohesie replied:
There's status effects that don't disappear based on time.
Check for examples in other effects, and the status effects variables and their documentation.

cacogen replied:
I know how to add a status effect that doesn't disappear based on time. 

The issue is as I've described above. 

I tried for hours to get this to work using on_mob_metabolize() and on_mob_end_metabolize() (the way reagents like methamphetamine add and remove their movespeed modifiers) before oranges suggested using a timed status effect. Then it finally worked, which was a huge relief.

I just tried using on_mob_metabolize() and on_mob_end_metabolize() again with the status effect. I set the status effect to not have a duration. It appeared to not be applied at all.

When I was trying to get it to work before the status effect, I used debug messages to discover that the things being applied with on_mob_metabolize() were immediately being removed by on_mob_end_metabolize().

No amount of changing the metabolisation rate helped. I tried using methamphetamine's metabolisation rate, because it applies a movement speed modifier that sticks. No luck. I would've tried other metabolisation rates too.



You must know from personal experience how tedious it is to compile the code, wait for the world to finish setting up and get into testing it.

cacogen replied:
I know how to add a status effect that doesn't disappear based on time. 

The issue is as I've described above. 

I tried for hours to get this to work using on_mob_metabolize() and on_mob_end_metabolize() (the way reagents like methamphetamine add and remove their movespeed modifiers) before oranges suggested using a timed status effect. Then it finally worked, which was a huge relief.

I just tried using on_mob_metabolize() and on_mob_end_metabolize() again with the status effect. I set the status effect to not have a duration. It appeared to not be applied at all.

When I was trying to get it to work before the status effect, I used debug messages to discover that the things being applied with on_mob_metabolize() were immediately being removed by on_mob_end_metabolize().

No amount of changing the metabolisation rate helped. I tried using methamphetamine's metabolisation rate, because it applies a movement speed modifier that sticks. No luck. I would've tried other metabolisation rates too.



You must know from personal experience how tedious it is to compile the code, wait for the world to finish setting up and get into testing it.

cacogen replied:
I tried for hours to get it to work with on_mob_metabolise() and on_mob_end_metabolise() the way you're asking for now. It was incredibly frustrating. @optimumtact suggested using a time status effect and it worked. I have no idea how to get it to work the way you're asking for and I don't want to go through trying to get it to work that way again.

optimumtact replied:
@Rohesie the way cigs microdose reagents mean it's constantly added and removed

Rohesie replied:
I see. If that's the case this will be hacky one way or another, until a better system is cooked.

cacogen committed:
Alters comments added in drug_effects.dm

- Simplifies BLOODCULT_EYE check
- Removes return = TRUE thing I don't understand

Rohesie commented on code/modules/reagents/chemistry/reagents/drug_reagents.dm:
```diff
@@ -34,6 +34,31 @@
+	if(DT_PROB(4, delta_time) && M.buckled && M.body_position != LYING_DOWN && !M.IsParalyzed()) //chance to be couchlocked if sitting
+		to_chat(M, "<span class='warning'>It's too comfy to move...</span>")
+		M.Paralyze(10 SECONDS)
+	..()
```
```suggestion
	return ..()
```
Preserve the parent return unless you have a specific reason not to.

Once this change is made and you've tested this locally, if you haven't since your last change, set the PR again as ready for review and it should be good to merge.

cacogen committed:
Update code/modules/reagents/chemistry/reagents/drug_reagents.dm

Co-authored-by: Rohesie <rohesie@gmail.com>

Rohesie commented on code/datums/status_effects/drug_effects.dm:
```diff
@@ -58,3 +58,39 @@
+	var/original_eye_color
+
+/datum/status_effect/stoned/on_apply()
+	if(ishuman(owner))
+		var/mob/living/carbon/human/human_owner = owner
+		original_eye_color = human_owner.eye_color
+		human_owner.add_movespeed_modifier(/datum/movespeed_modifier/reagent/cannabis) //slows you down
+		human_owner.eye_color = BLOODCULT_EYE //makes cult eyes less obvious
+		human_owner.update_body() //updates eye color
+		ADD_TRAIT(human_owner, TRAIT_BLOODSHOT_EYES, type) //dilates blood vessels in eyes
+		ADD_TRAIT(human_owner, TRAIT_CLUMSY, type) //impairs motor coordination
+		SEND_SIGNAL(human_owner, COMSIG_ADD_MOOD_EVENT, "stoned", /datum/mood_event/stoned) //improves mood
+		human_owner.sound_environment_override = SOUND_ENVIRONMENT_DRUGGED //not realistic but very immersive
+	return TRUE
```
```suggestion
	if(!ishuman(owner))
		return FALSE
	var/mob/living/carbon/human/human_owner = owner
	original_eye_color = human_owner.eye_color
	human_owner.add_movespeed_modifier(/datum/movespeed_modifier/reagent/cannabis) //slows you down
	human_owner.eye_color = BLOODCULT_EYE //makes cult eyes less obvious
	human_owner.update_body() //updates eye color
	ADD_TRAIT(human_owner, TRAIT_BLOODSHOT_EYES, type) //dilates blood vessels in eyes
	ADD_TRAIT(human_owner, TRAIT_CLUMSY, type) //impairs motor coordination
	SEND_SIGNAL(human_owner, COMSIG_ADD_MOOD_EVENT, "stoned", /datum/mood_event/stoned) //improves mood
	human_owner.sound_environment_override = SOUND_ENVIRONMENT_DRUGGED //not realistic but very immersive
	return TRUE
```
* Early return and return false if the owner is not human. We don't want to keep the trait then.

Rohesie commented on code/datums/status_effects/drug_effects.dm:
```diff
@@ -58,3 +58,39 @@
+	return TRUE
+
+/datum/status_effect/stoned/on_remove()
+	if(ishuman(owner))
+		var/mob/living/carbon/human/human_owner = owner
+		human_owner.remove_movespeed_modifier(/datum/movespeed_modifier/reagent/cannabis)
+		human_owner.eye_color = original_eye_color
+		human_owner.update_body()
+		REMOVE_TRAIT(human_owner, TRAIT_BLOODSHOT_EYES, type)
+		REMOVE_TRAIT(human_owner, TRAIT_CLUMSY, type)
+		SEND_SIGNAL(human_owner, COMSIG_CLEAR_MOOD_EVENT, "stoned")
+		human_owner.sound_environment_override = SOUND_ENVIRONMENT_NONE
```
```suggestion
	if(!ishuman(owner))
			stack_trace("[type] status effect being removed from non-human owner: [owner ? owner.type : "null owner"]")
	var/mob/living/carbon/human/human_owner = owner
	human_owner.remove_movespeed_modifier(/datum/movespeed_modifier/reagent/cannabis)
	human_owner.eye_color = original_eye_color
	human_owner.update_body()
	REMOVE_TRAIT(human_owner, TRAIT_BLOODSHOT_EYES, type)
	REMOVE_TRAIT(human_owner, TRAIT_CLUMSY, type)
	SEND_SIGNAL(human_owner, COMSIG_CLEAR_MOOD_EVENT, "stoned")
	human_owner.sound_environment_override = SOUND_ENVIRONMENT_NONE
```
* Make use of early returns.
* Avoid silent sanity checks, make them scream.

Rohesie commented:
After those changes are done test your code locally.
If everything runs as you expect it and you find no problems, set it again as ready for review.

cacogen committed:
Early return, proper debug message, return falsey if the owner is not human

Co-authored-by: Rohesie <rohesie@gmail.com>

cacogen committed:
Early return, makes sanity check scream

Co-authored-by: Rohesie <rohesie@gmail.com>

cacogen committed:
Fixes stack_trace indentation

- Couple of last minute tweaks

Tumesuo commented:
Hello, this weed is unrealistic. 
regards, professional weed toker (it's slang for smoking weed)

cacogen commented:
You're welcome to improve upon it

MrDoomBringer commented:
Virtual weed could never meet the high standards that we in the professional weed toker community uphold in real life
